### DISCUSSION TITLE
Can juicefs recover from lost metadata? E.g. if redis is gone?

### DISCUSSION THREAD

**Discussion by ghettogeets** [juicedata/juicefs#131] [Q&A] [answered]
Hi,

we are evaluating juicefs as a storage backend for a video-transcoding system. First tests worked out nicely!

But I am concerned what happens if we lost data on the redis-db.
Is there a way to rebuild solely from the data on the bucket?

**Comment by xiaogaozi** [comment, marked as answer]
> Is there a way to rebuild solely from the data on the bucket?

No. The metadata in Redis is very important, if you loose these data, you couldn't restore the file system.

We recommend read Redis [official document](https://redis.io/topics/persistence) about how to persistent data first. If you run your service on public cloud, use the Redis service hosted by the provider is a good choice.

**Comment by ghettogeets** [comment]
Hello and thank you very much for your answer.

Here's another thought I came across:
Is there a way to handle backups?

To get a full and consistent backup we would have to backup redis and the storage bucket at the same time. Is that right?

Thanks in advance
Alex

**Reply by xiaogaozi** [reply]
> To get a full and consistent backup we would have to backup redis and the storage bucket at the same time. Is that right?

In theory, yes. But object storage provided by public cloud has very high data durability (e.g. S3 provides 99.999999999% durability), just backup Redis is enough.

**Reply by sloonz** [reply]
Let's suppose that I lose redis metadata.  I restore a week-old redis metadata backup, whereas the data in S3 is still the most recent one. Will it work ?

Here is my worry : 

1. I edit the file `foo`, set it to "hello"
2. I backup the metadata
3. I edit the file `foo`, set it to "bar". The initial data, "hello", is removed from S3
4. I lose my Redis instance, and restore the backup at (2)
5. At this point, `foo` is neither at "hello" (because it was overwritten by the most recent write) and neither at "bar" (because metadata is pointing at ancient data)

It is something that can happen ?

> S3 provides 99.999999999% durability

That is a theoretical value that doesn’t takes into account mundane failure modes such as "my account was suspended by amazon for payment failure" or "i fat-fingered my `s3cmd ls -r` into `s3cmd rm -r`". I don’t think "just backup redis" is a good answer to the question of "how do i backup my data" ?

**Comment by Suave** [comment]
@ghettogeets Here's another way to backup metadata, check the detail at https://juicefs.com/docs/community/metadata_dump_load

You could dump metadata to a JSON file, you could load it into another metadata engine. Also here's an auto backup policy for https://juicefs.com/docs/community/metadata_dump_load#automatic-backup
